src/include/miopen/binary_serialization.hpp

```diff
@@ -0,0 +1,229 @@
+/*******************************************************************************
+ *
+ * MIT License
+ *
+ * Copyright (c) 2022 Advanced Micro Devices, Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ *
+ *******************************************************************************/
+
+#pragma once
+
+#include <miopen/errors.hpp>
+#include <miopen/type_traits.hpp>
+
+#include <cstring>
+#include <type_traits>
+
+namespace miopen {
+
+namespace detail {
+class BinarySerializationRelatedStream
+{
+protected:
+    template <class Type, class = void>
+    struct SerializationCriteria : std::false_type
+    {
+    };
+
+    template <class Type>
+    struct SerializationCriteria<
+        Type,
+        std::enable_if_t<std::is_pod<Type>{} && !std::is_pointer<Type>{} && !std::is_empty<Type>{},
+                         void>> : std::true_type
+    {
+    };
+};
+} // namespace detail
+
+class BinarySerializationSizeStream final : detail::BinarySerializationRelatedStream
+{
+public:
+    BinarySerializationSizeStream() = default;
+
+    bool IsSerializing() const { return false; }
+    bool IsDeserializing() const { return false; }
+
+    template <class Type, std::enable_if_t<SerializationCriteria<Type>{}, bool> = true>
+    BinarySerializationSizeStream& operator<<(const Type&)
+    {
+        size += sizeof(Type);
+        return *this;
+    }
+
+    std::size_t GetSize() const { return size; }
+
+private:
+    std::size_t size = 0;
+};
+
+class BinarySerializationStream final : detail::BinarySerializationRelatedStream
+{
+public:
+    bool IsSerializing() const { return true; }
+    bool IsDeserializing() const { return false; }
+
+    BinarySerializationStream(char* buffer) : position(buffer) {}
+
+    template <class Type, std::enable_if_t<SerializationCriteria<Type>{}, bool> = true>
+    BinarySerializationStream& operator<<(const Type& value)
+    {
+        std::memcpy(position, &value, sizeof(Type));
```

Comment:
This is not portable. We should not be creating our own binary serialization(which could expose security vulnerabilities as well).

Instead, we should use msgpack to serialize to binaries. If we use nlohmann/json, It also can serialize STL containers, and it has an easy interface to extend to your own data structures so we dont need to write our own serialization framework.

Also, msgpack makes it easier for us to read these buffers for debugging purposes.

In #1290, I added nlohmann/json as a dependency but it is header-only so if you dont like a dependency you can just copy the single header over to our repo.